src/provider/provider_level_zero.c

```diff
@@ -930,6 +982,162 @@ static umf_result_t ze_memory_provider_get_allocation_properties_size(
     return UMF_RESULT_ERROR_INVALID_ARGUMENT;
 }
 
+struct ze_memory_provider_resident_device_change_data {
+    bool is_adding;
+    ze_device_handle_t peer_device;
+    ze_memory_provider_t *source_memory_provider;
+    uint32_t success_changes;
+    uint32_t failed_changes;
+};
+
+static int ze_memory_provider_resident_device_change_helper(uintptr_t key,
+                                                            void *value,
+                                                            void *privdata) {
+    struct ze_memory_provider_resident_device_change_data *change_data =
+        privdata;
+    tracker_alloc_info_t *info = value;
+    if (info->props.provider->provider_priv !=
+        (void *)change_data->source_memory_provider) {
+        LOG_DEBUG("ze_memory_provider_resident_device_change found not our "
+                  "pointer %p",
+                  (void *)key);
+        return 0;
+    }
+
+    assert(info->props.base == (void *)key);
+
+    ze_result_t result;
+    if (change_data->is_adding) {
+        result = g_ze_ops.zeContextMakeMemoryResident(
+            change_data->source_memory_provider->context,
+            change_data->peer_device, info->props.base, info->props.base_size);
+    } else {
+        result = ZE_RESULT_SUCCESS;
+        // TODO: currently not implemented call evict here
+    }
+
+    if (result != ZE_RESULT_SUCCESS) {
+        LOG_ERR(
+            "ze_memory_provider_resident_device_change found our pointer "
+            "%p but failed to make it resident on device: %p due to err: %d",
+            (void *)key, (void *)change_data->peer_device, result);
+        ++change_data->failed_changes;
+        return 1;
+    }
+
+    LOG_DEBUG("ze_memory_provider_resident_device_change found our pointer %p "
+              "and made it resident on device: %p",
+              (void *)key, (void *)change_data->peer_device);
+    ++change_data->success_changes;
+    return 0;
+}
+
+umf_result_t umfLevelZeroMemoryProviderResidentDeviceChange(
```

Comment:
I would prefer to have add and remove function - i recall pbalcer already suggested it too. 

I will split it to 3 functions - add, delete, and helper function to iterate over existing allocations (so we limit code duplication)